smack-websocket/src/main/java/org/jivesoftware/smack/websocket/XmppWebsocketTransportModule.java

```diff
@@ -0,0 +1,308 @@
+        }
+    }
+
+    private DiscoveredWebsocketEndpoints discoveredWebsocketEndpoints;
+
+    /**
+     * Transport class for {@link ModularXmppClientToServerConnectionModule}'s websocket implementation.
+     */
+    public final class XmppWebsocketTransport extends XmppClientToServerTransport {
+
+        protected XmppWebsocketTransport(ModularXmppClientToServerConnectionInternal connectionInternal) {
+            super(connectionInternal);
+        }
+
+        @Override
+        protected void resetDiscoveredConnectionEndpoints() {
+            discoveredWebsocketEndpoints = null;
+        }
+
+        @Override
+        protected List<SmackFuture<LookupConnectionEndpointsResult, Exception>> lookupConnectionEndpoints() {
+            // Assert that there are no stale discovered endpoints prior performing the lookup.
+            assert discoveredWebsocketEndpoints == null;
+
+            List<SmackFuture<LookupConnectionEndpointsResult, Exception>> futures = new ArrayList<>(2);
+
+            InternalSmackFuture<LookupConnectionEndpointsResult, Exception> websocketEndpointsLookupFuture = new InternalSmackFuture<>();
+
+            connectionInternal.asyncGo(() -> {
+
+                WebsocketRemoteConnectionEndpoint providedEndpoint = null;
+
+                // Check if there is a websocket endpoint already configured.
+                if (moduleDescriptor.getUri() != null) {
+                    providedEndpoint = new WebsocketRemoteConnectionEndpoint(moduleDescriptor.getUri());
+                }
+
+                if (!moduleDescriptor.isWebsocketEndpointDiscoveryEnabled()) {
+                    // If discovery is disabled, assert that the provided endpoint isn't null.
+                    assert providedEndpoint != null;
+
+                    if ((providedEndpoint.isSecureEndpoint() &&
+                            connectionInternal.connection.getConfiguration().getSecurityMode().equals(SecurityMode.disabled))
+                            || (!providedEndpoint.isSecureEndpoint() &&
+                                    connectionInternal.connection.getConfiguration().getSecurityMode().equals(SecurityMode.required))) {
+                        throw new IllegalStateException("Endpoint provided does not satisfy the security mode configuration");
+                    }
+
+                    // Generate Result for manual endpoint.
+                    Result manualResult = new Result(Arrays.asList(providedEndpoint), null);
+
+                    LookupConnectionEndpointsResult endpointsResult = new DiscoveredWebsocketEndpoints(manualResult);
+
+                    websocketEndpointsLookupFuture.setResult(endpointsResult);
+                } else {
+                    // Fetch remote endpoints.
+                    Result xep0156result = WebsocketRemoteConnectionEndpointFinder.find(connectionInternal);
+
+                    List<WebsocketRemoteConnectionEndpoint> endpoints = new ArrayList<>();
```

Comment:
Size the array list accordingly. First get the discovered remote connection endpoints out of the xep0156 result, and then create the list of endpoints with the correct size, i.e. the number of discovered endpoints. You don't have to take into account that there is potentially a `providedEndpoint`, as this would be to much code for to little gain.